Construct VideoFrame from Canvas & OffscreenCanvas

mstyura opened the pull request:
#### d77809d46e44e2748bb982d6365fd13b17acc10d
<pre>
Construct VideoFrame from Canvas &amp; OffscreenCanvas

<a href="https://bugs.webkit.org/show_bug.cgi?id=300579">https://bugs.webkit.org/show_bug.cgi?id=300579</a>

Reviewed by Kimmo Kinnunen.

This add two simple tests to measure performance of construction of
VideoFrame from Canvas &amp; OffscreenCanvas. The ineffective implementation
of VideoFrame constructor does copy between processes and transformation
of pixels and will be slow in benchmark.

* PerformanceTests/Canvas/videoFrameFromCanvas.html: Added.
* PerformanceTests/Canvas/videoFrameFromOffscreenCanvas.html: Added.

Canonical link: <a href="https://commits.webkit.org/302974@main">https://commits.webkit.org/302974@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e5ad2070a790231352ac3d9d65bae8ba18539b68

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130624 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2895 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41579 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138044 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82246 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2907 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2787 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99516 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/67381 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133571 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2110 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116964 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80223 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2031 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81301 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110606 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35600 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140523 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2685 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2435 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108020 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2729 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113308 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107953 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27503 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2072 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31735 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/55667 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2755 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2574 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2776 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2681 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on PerformanceTests/Canvas/drawImageSourceVideoFrameCanvasWebGL.html:
```diff
@@ -0,0 +1,29 @@
+<!DOCTYPE html>
```
Could you:
* call this file drawImageSourceVideoFrameCanvasWebGL.html?
* add 2D canvas that gets drawImage to use the video frame before closing it. It's always good to test something that actually does something..
* add few iterations (in case it's fast. in case it's very slow, iterations could be 1)
* maybe use a bit weirder dimensions, maybe non-power of two, just in case.

mstyura replied:
Thanks a lot for the review! I've adjusted test according to your comments (I hope I understand it correctly).

> add few iterations (in case it's fast. in case it's very slow, iterations could be 1)
> maybe use a bit weirder dimensions, maybe non-power of two, just in case.

I preferred to keep big canvas size without nested iterations over the same unchanged canvas. I mean that each iteration has the canvas cleared, new frame created, and painted over different canvas. I such case I expect there will be no internal caching for unchanged data. So I expect any ineffective copy will become bottleneck sooner.

The additional problem with big canvas size I've noticed, is that it seems like both Safari and Chrome release memory non-immediately, which for canvas sizes > 50 Megapixels on my Mac leads to a page crash due GPU process memory consumption. Firefox is very stable at memory consumption on modified benchmark.

kkinnunen-apple commented on PerformanceTests/Canvas/videoFrameFromOffscreenCanvas.html:
```diff
@@ -0,0 +1,21 @@
+<!DOCTYPE html>
```
Same comments, drawImageSourceVideoFrameOffscreenCanvasWebGL.html

kkinnunen-apple reviewed (commented):
Thank you, great! Small requests, still. Sorry, I didn't see this earlier. (If you want quicker reviews, feel free to ping in Slack)

kkinnunen-apple reviewed (approved):
Thank you, looks good.
(WebKit behaviour looks bad, need to fix it)

mstyura committed:
Construct VideoFrame from Canvas & OffscreenCanvas

https://bugs.webkit.org/show_bug.cgi?id=300579

Reviewed by Kimmo Kinnunen.

This add two simple tests to measure performance of construction of
VideoFrame from Canvas & OffscreenCanvas. The ineffective implementation
of VideoFrame constructor does copy between processes and transformation
of pixels and will be slow in benchmark.

* PerformanceTests/Canvas/videoFrameFromCanvas.html: Added.
* PerformanceTests/Canvas/videoFrameFromOffscreenCanvas.html: Added.

Canonical link: https://commits.webkit.org/302974@main
